lib/IRGen/IRGenSIL.cpp

```diff
@@ -2693,28 +2821,45 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
   auto *deleteStatusFn = IGM.getTF_DeleteStatusFn();
   Builder.CreateCall(deleteStatusFn, {status});
 
-  for (unsigned resultIdx = 0; resultIdx < i->getNumResults(); ++resultIdx) {
-    auto resultIdxVal = llvm::ConstantInt::get(IGM.Int32Ty, resultIdx);
-    auto resultAddr = Builder.CreateInBoundsGEP(returnValues.getAddress(),
-                                                resultIdxVal);
-    auto cTensorHandle = Builder.CreateLoad(resultAddr,
-                                            IGM.getPointerAlignment());
-
-    // Wrap `cTensorHandle` into a _AnyTensorHandle object, and get an untyped
-    // pointer to the _AnyTensorHandle.
-    auto *createHandleFn = IGM.getTFC_CreateTensorHandleFromCFn();
-    llvm::Value *tensorHandle = Builder.CreateCall(createHandleFn,
-                                                   {cTensorHandle});
-
-    // Cast to a pointer of the expected result type (for example,
-    // TensorHandle<Float>).
-    SILValue result = i->getResults()[resultIdx];
-    tensorHandle = Builder.CreateBitCast(
-        tensorHandle, IGM.getStorageType(result->getType()));
+  if (outParameterAddress.isValid()) {
+    // If we have an out parameter, store the results there.
+    assert(i->getNumResults() == 0 &&
+           "can't handle direct and indirect results at the same time");
+    auto *initTensorGroupFn = IGM.getTFC_InitTensorGroupFn();
+    auto *outParameterAddressOpaque = Builder.CreateBitCast(
+        outParameterAddress.getAddress(), IGM.OpaquePtrTy);
+    auto *returnValuesAddressUntyped = Builder.CreateBitCast(
+        returnValuesAddress, IGM.Int8PtrTy);
+    Builder.CreateCall(initTensorGroupFn,
+                       {outParameterAddressOpaque,
+                        returnValuesAddressUntyped,
+                        outParameterTypeMetadata,
+                        outParameterTensorGroupWitnessTable});
+  } else {
```

Comment:
Since this sits directly above a return, would it be nicer to do an early return in the if-branch and out-scope the else-branch?

Reply:
I do not think early return makes sense here. It's two "equally successful" cases, not one case where we want to bail out and another case where we want to keep doing a bunch more processing.